### ISSUE TITLE
Support for ghc-9.2.*

### ISSUE THREAD

**Issue by jneira** [haskell/haskell-language-server#2179]
* Similar issue to #297
* There is a work in progress by @fendor started with #2128 although it still does not add initial support, only preparing it (maybe @fendor could tell us more about)

--------------------------------

## State of support

* The core functionality of haskell-language-server is supported via ghcide. 
* Plugins and formatters needs to be updated to make it work with ghc-9.0.1 so they are not included until then.
* So you should consider use other ghc version only if you are specially interested in those plugins o formatters 
* The plan is include a prebuilt executable for ghc-9.2.1 in the next hls release (1.6.0)

## Has ghc-9.2.1 support
- [x] `ghcide`
- [x] `haskell-language-server`
- [ ] `hls-haddock-comments-plugin`
- [x] `hls-hlint-plugin`
- [x] `hls-pragmas-plugin`
- [ ] `hls-retrie-plugin` without test, not included via flags
- [x] `hls-eval-plugin`
- [x] `hls-module-name-plugin`: it was not being tested in ci: see #2587. Doesnt compile for ghc-9.2.1. Fixed with #2594
- [x] `hls-class-plugin`
- [x] `hls-refine-imports-plugin`: #2601
- [ ] `hls-splice-plugin`
- [ ] `hls-tactic-plugin` 
- [x] `hls-explicit-imports-plugin`
- [x] `hls-call-hierarchy-plugin` it is being tested so it works but it is not included in hls via flags: see #2587
- [x] `hls-alternate-number-format-plugin`
- [x] `hls-qualify-imported-names-plugin`: #2600 
 
### Formatters
- [x] `hls-floskell-plugin`
- [x] `hls-ormolu-plugin`
- [x] `hls-fourmolu-plugin`
- [ ] `hls-brittany-plugin`
- [x] `hls-stylish-haskell-plugin`

### Build system and CI
- [x] included in github release
  - [x] `.github/workflows/build.yml` 
- [ ] hackage release with support for ghc-9.2.1: #2606 
  - [ ] `.github/workflows/hackage.yml` 
- [x] `stack-9.2.1.yaml`, added but not working
  - [x] `.circleci/config.yml`
- [x] `cabal-ghc921.project` - Used for disabling the not yet working plugins and fetching upstream patches
- [x]  nix support: with #2621 
  - [x] ./`configuration-ghc-921.nix`
  - [x] update `.github/workflows/nix.yml` 
- [x] `.github/workflows/test.yml` - Uses the custom cabal.project and disables tests for non-working plugins. Needs to be updated when more plugins get ghc-9.2.1 support

### Dependencies which needs updates and hackage release
- TODO: track all dependencies with issues(prs taking as base the `allow-newer` field of `cabal-ghc921.project`

**Comment by fendor** [comment]
The MR #2128 adds most import statements and functions compatible with the GHC API 9.2 in a backward compatible manner. With I think roughly 3 exceptions (which require some small changes), bringing basic support for ghc-9.2 is a matter of adding the appropriate imports, once retrie supports ghc 9.2.

**Comment by Dessix** [comment]
Looks like the next portion of this is dependent on [support for 9.2 in retrie](https://github.com/facebookincubator/retrie/issues/36), which has a WIP that needs carried over the finish-line.

**Comment by mouse07410** [comment]
How near are we to releasing 1.5.0, and would it include GHC-9.2.1 support?

**Comment by jneira** [comment]
Hi, 1.5.0 will be released soon but i am afraid it will not include ghc-9.2.1 for sure, there is still lot of work to be done here

**Comment by mouse07410** [comment]
> I am afraid it will not include ghc-9.2.1 for sure, there is still lot of work to be done here

Maybe at least include partial support??? (_Assuming it can be done and makes sense_)

**Comment by jneira** [comment]
> 
> 
> > I am afraid it will not include ghc-9.2.1 for sure, there is still lot of work to be done here
> 
> Maybe at least include partial support??? (_Assuming it can be done and makes sense_)

I am afraid that this is blocked on a [ghcide change](https://github.com/haskell/haskell-language-server/issues/2179#issuecomment-932487785) and ghcide is the core of hls so we cant release it even partially.

**Comment by Ailrun** [comment]
Related issue: nikita-volkov/ptr-poker#6 (which is a dependency of jsonify, which is in turn of opentelemetry-extra)

**Comment by Dessix** [comment]
[ghc-exactprint-1.3.0](https://hackage.haskell.org/package/ghc-exactprint-1.3.0) has been released - that was the portion that was blocking retrie itself; the next step is to update retrie to use it, and then it can be used for HLS 9.2.

**Comment by Ailrun** [comment]
@Dessix The branch is actually merged now. It's now the problem of release (However, even after that, we still have opentelemetry-extra issue)

**Comment by Dessix** [comment]
Looks like the retrie release [went through](/facebookincubator/retrie/issues/36#issuecomment-993366995).

[`ptr-poker #7`](/nikita-volkov/ptr-poker/pull/7) is also released. What can we do to move this forward into `jsonify` and subsequently `opentelemetry-extra`?

**Comment by Ailrun** [comment]
Unfortunately, there was a mistake in the PR and fix committed afterwards: https://github.com/nikita-volkov/ptr-poker/pull/8

**Comment by pepeiborra** [comment]
We don't really depend on `opentelemetry-extra`, only on `opentelemetry`. Does that help?

**Comment by Ailrun** [comment]
Huh? then Why do I get issues from a transitive dependency of `opentelemetry-extra`? That's weird... let me check

**Comment by pepeiborra** [comment]
`opentelemetry-extra` provides tools to analyse eventlogs in Tracy, Chrome and other open telemetry viewers. The flake builds it because I use it.

**Comment by Ailrun** [comment]
GHC 9.2 support of GHCIDE (probably not yet exhaustive):
### Main
- [ ] Aeson version update #2477  
  Aeson 2.0.2.0 supports GHC 9.2 (haskell/aeson#885)
- [ ] New release of [`constraints-extras`](https://github.com/obsidiansystems/constraints-extras) (or we can use their HEAD)
- [ ] Support of primitive-unlifted (haskell-primitive/primitive-unlifted#28)
- [ ] Support of th-extras (mokus0/th-extras#15)
- [ ] Support of opentelemetry (https://github.com/ethercrow/opentelemetry-haskell CI gives errors for GHC 9.2)
### Testing
- [ ] Support of record-dot-preprocessing (ndmitchell/record-dot-preprocessor#43)

**Comment by pepeiborra** [comment]
`opentelemetry` installs without issues for me:
```
cabal install opentelemetry --overwrite-policy=always -w ghc-9.2 
Resolving dependencies...
Build profile: -w ghc-9.2.1 -O1
In order, the following will be built (use -v for more details):
 - ghc-trace-events-0.1.2.4 (lib) (requires build)
 - opentelemetry-0.7.0 (lib) (requires build)
Starting     ghc-trace-events-0.1.2.4 (lib)
Building     ghc-trace-events-0.1.2.4 (lib)
Installing   ghc-trace-events-0.1.2.4 (lib)
Completed    ghc-trace-events-0.1.2.4 (lib)
Starting     opentelemetry-0.7.0 (lib)
Building     opentelemetry-0.7.0 (lib)
Installing   opentelemetry-0.7.0 (lib)
Completed    opentelemetry-0.7.0 (lib)
```

**Comment by pepeiborra** [comment]
WIP branch here: https://github.com/haskell/haskell-language-server/compare/ghc-9.2?expand=1

**Comment by Ailrun** [comment]
Awesome, let's wait how it goes in other repositories...

**Comment by wz1000** [comment]
We can use `head.hackage` to obtain a build plan. It already has patches for [th-extras](https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/patches/th-extras-0.0.0.5.patch), [constraints-extras](https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/patches/constraints-extras-0.3.2.0.patch) and [primitive-unlifted](https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/patches/primitive-unlifted-0.1.3.0.patch)

**Comment by pepeiborra** [comment]
> We can use `head.hackage` to obtain a build plan. It already has patches for [th-extras](https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/patches/th-extras-0.0.0.5.patch), [constraints-extras](https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/patches/constraints-extras-0.3.2.0.patch) and [primitive-unlifted](https://gitlab.haskell.org/ghc/head.hackage/-/blob/master/patches/primitive-unlifted-0.1.3.0.patch)

It has a patch for `primitive-unlifted-0.1.3.0` but it looks like we need a newer version:
```
cabal: Could not resolve dependencies:
[__0] trying: ghcide-1.5.0.1 (user goal)
[__1] trying: stm-containers-1.2 (dependency of ghcide)
[__2] trying: stm-hamt-1.2.0.7 (dependency of stm-containers)
[__3] trying: primitive-extras-0.10.1.2 (dependency of stm-hamt)
[__4] next goal: primitive-unlifted (dependency of primitive-extras)
[__4] rejecting: primitive-unlifted-1.0.0.0, primitive-unlifted-0.1.3.1
(constraint from project config
/Users/pepeiborra/scratch/ide/cabal-ghc921.project requires ==0.1.3.0)
[__4] rejecting: primitive-unlifted-0.1.3.0 (conflict: primitive-extras =>
primitive-unlifted>=1.0 && <1.1)
```

**Comment by Bodigrim** [comment]
Because of https://github.com/haskell-primitive/primitive-unlifted/issues/28 you need the following incantation:
```cabal
allow-newer:
  primitive-unlifted:base,
allow-older:
  primitive-extras:primitive-unlifted
constraints:
  primitive-unlifted < 1
```
If the issue with GHC 9.2 support in `primitive-unlifted` gets stalled, Hackage trustees can make revisions to avoid `allow-newer` / `allow-older`.

**Comment by Bodigrim** [comment]
I believe `primitive-{extras,unlifted}` tweaks are no longer needed now.

**Comment by Dessix** [comment]
#2503 has been merged, as have `primitive-unlifted` and `Aeson` support for 9.2.1; What's the next blocker, or are we just waiting for the next release?

**Comment by pepeiborra** [comment]
HEAD already has experimental support for GHC 9.2 - just waiting for a release now.

**Comment by jneira** [comment]
I've updated the issue description with a detailed check list tracking ghc-9.2.1 support

**Comment by lilnasy** [comment]
If there are beginner friendly tasks for 9.2 support, I'd love to help out.

**Comment by jneira** [comment]
> If there are beginner friendly tasks for 9.2 support, I'd love to help out.

Many thanks for the offer. Well there are some plugins without 9.2.1 support. Some of thems need updates in the underlying library so they are blocked. The plugins which could be adapted would be:
* hls-haddock-comments-plugin: maybe needs some extensive changes dur to eaxctprint changes in 9.2.1 (not sure //cc @berberman )
* hls-eval-plugin: complex codebase, may be difficult (//cc @tittoassini)
* hls-class-plugin: maybe the easier to tackle? (//cc @Ailrun)
* hls-tactics-plugin: complex codebase, recently added to 9.0 (//cc @isovector)

**Comment by mouse07410** [comment]
Somewhat out of topic: what about 9.0.2? Stackage moved nightly to 9.0.2, but HLS is still 9.0.1-only.

**Comment by jneira** [comment]
> Somewhat out of topic: what about 9.0.2? Stackage moved nightly to 9.0.2, but HLS is still 9.0.1-only.

Support was tracked with https://github.com/haskell/haskell-language-server/issues/2541 and added with https://github.com/haskell/haskell-language-server/pull/2567 so the next release will include it

**Comment by TeofilC** [comment]
hlint-3.4 has just been released with support for ghc-9.2 https://hackage.haskell.org/package/hlint-3.4/changelog

**Comment by konn** [comment]
GHC 9.2.3 has just been [released](https://www.haskell.org/ghc/download_ghc_9_2_3.html), so it might be the time to tackle it.

**Comment by Dessix** [comment]
#2732 (OverloadedRecordDot support) may be worth adding to the checklist, as the IDE gets somewhat confused with 9.2's new record syntax otherwise.

**Comment by mouse07410** [comment]
Any work on GHC-9.2.3 yet? It shouldn't be too different from 9.2.2 from the HLS point of view, I'd imagine?

**Comment by maerten** [comment]
There's a PR for 9.2.3: https://github.com/haskell/haskell-language-server/pull/2936

**Comment by jneira** [comment]
Sorry this issue is outdated and i dont have time to keep up to date, please open a new one if you want to track support for newer versions of ghc-9.2
